### PR TITLE
[feature] plane-wave apodization in the elevation direction

### PR DESCRIPTION
**Pull request by charlesbmi**
so that for a 2D-array-probe we can use:

```
CombinedApodization(
    apodizations=[PlaneWaveTransmitApodization(array_bounds, apodization_dimension="azimuth"), PlaneWaveTransmitApodization(array_bounds, apodization_dimension="elevation")],
)
```

Added some unit tests that pass.

### CODE REVIEW

**Comment by charlesbmi** [comment]
I tested this visually as well by running the following script:

```
import numpy as np
import matplotlib.pyplot as plt
from vbeam.scan import linear_scan
from vbeam.core import ElementGeometry, WaveData
from vbeam.apodization.plane_wave import PlaneWaveTransmitApodization
from vbeam.apodization.window import Hanning

# Create scans in both y-z and x-z planes
y_points = np.linspace(-5e-3, 5e-3, 100)  # y coordinates in meters
x_points = np.linspace(-5e-3, 5e-3, 100)  # x coordinates in meters
z_points = np.linspace(0, 10e-3, 100)  # z coordinates in meters
yz_scan = linear_scan(np.array([0.0]), y_points, z_points)  # for elevation
xz_scan = linear_scan(x_points, z_points)  # for azimuth

# Define array bounds (same physical size for both directions)
array_size = 2.5e-3  # 2.5mm on each side
array_bounds_elevation = (
    np.array([0, -array_size / 2, 0]),  # [x, y, z] left/bottom
    np.array([0, array_size / 2, 0]),  # [x, y, z] right/top
)
array_bounds_azimuth = (
    np.array([-array_size, 0, 0]),  # [x, y, z] left
    np.array([array_size, 0, 0]),  # [x, y, z] right
)

# Create transmit element at center of array
transmit_element = ElementGeometry(position=np.array([0, 0, 0]))

# Create both apodization instances
apodization_elevation = PlaneWaveTransmitApodization(
    array_bounds=array_bounds_elevation,
    # window=Hanning(),
    apodization_dimension="elevation",
)

apodization_azimuth = PlaneWaveTransmitApodization(
    array_bounds=array_bounds_azimuth,
    # window=Hanning(),
    apodization_dimension="azimuth",
)

# Test different angles
angles = [np.pi / 6, 0, -np.pi / 6]  # 30°, 0°, -30°

# Create figure with two rows
fig, axes = plt.subplots(2, len(angles), figsize=(15, 8))

for col, angle in enumerate(angles):
    # Plot elevation (top row)
    wave_data_elevation = WaveData(azimuth=0, elevation=angle)
    im_elev = axes[0, col].imshow(
        apodization_elevation(
            sender=transmit_element,
            point_position=yz_scan.get_points(),
            receiver=transmit_element,
            wave_data=wave_data_elevation,
        )
        .reshape(100, 100)
        .T,
        extent=[y_points[0], y_points[-1], z_points[-1], z_points[0]],
        aspect="auto",
        vmin=0,
        vmax=1,
    )
    axes[0, col].set_xlabel("y [m]")
    if col == 0:
        axes[0, col].set_ylabel("z [m]")
    axes[0, col].set_title(f"Elevation angle: {angle*180/np.pi:.1f}°")
    plt.colorbar(im_elev, ax=axes[0, col])

    # Plot azimuth (bottom row)
    wave_data_azimuth = WaveData(azimuth=angle, elevation=0)
    im_azim = axes[1, col].imshow(
        apodization_azimuth(
            sender=transmit_element,
            point_position=xz_scan.get_points(),
            receiver=transmit_element,
            wave_data=wave_data_azimuth,
        )
        .reshape(100, 100)
        .T,
        extent=[x_points[0], x_points[-1], z_points[-1], z_points[0]],
        aspect="auto",
        vmin=0,
        vmax=1,
    )
    axes[1, col].set_xlabel("x [m]")
    if col == 0:
        axes[1, col].set_ylabel("z [m]")
    axes[1, col].set_title(f"Azimuth angle: {angle*180/np.pi:.1f}°")
    plt.colorbar(im_azim, ax=axes[1, col])

plt.suptitle("Elevation (top) vs Azimuth (bottom) Apodization")
plt.tight_layout()
plt.show()

```
Figure_1


and also:
```
import numpy as np
import matplotlib.pyplot as plt
from vbeam.scan import linear_scan
from vbeam.core import ElementGeometry, WaveData
from vbeam.apodization.plane_wave import PlaneWaveTransmitApodization
from vbeam.apodization.window import Hanning
from vbeam.apodization.combine import combine_apodizations
import nibabel as nib

# Create 3D scan points
x_points = np.linspace(-5e-3, 5e-3, 100)  # x coordinates in meters
y_points = np.linspace(-5e-3, 5e-3, 100)  # y coordinates in meters
z_points = np.linspace(0, 10e-3, 100)  # z coordinates in meters

# Create 2D scans for visualization
yz_scan = linear_scan(np.array([0.0]), y_points, z_points)  # for elevation
xz_scan = linear_scan(x_points, z_points)  # for azimuth (y=0 by default)

# Create full 3D scan for volume export
xyz_scan = linear_scan(x_points, y_points, z_points)

# Define array bounds
array_size = 2.5e-3  # 2.5mm on each side
array_elevation_size = 1e-3
array_bounds_elevation = (
    np.array([0, -array_elevation_size, 0]),  # [x, y, z] bottom
    np.array([0, array_elevation_size, 0]),  # [x, y, z] top
)
array_bounds_azimuth = (
    np.array([-array_size, 0, 0]),  # [x, y, z] left
    np.array([array_size, 0, 0]),  # [x, y, z] right
)

# Create transmit element at center of array
transmit_element = ElementGeometry(position=np.array([0, 0, 0]))

# Create both apodization instances
apodization_elevation = PlaneWaveTransmitApodization(
    array_bounds=array_bounds_elevation,
    # window=Hanning(),
    apodization_dimension="elevation",
)

apodization_azimuth = PlaneWaveTransmitApodization(
    array_bounds=array_bounds_azimuth,
    # window=Hanning(),
    apodization_dimension="azimuth",
)

# Combine the apodizations
combined_apodization = combine_apodizations(apodization_azimuth, apodization_elevation)

# Test different angles
angles = [np.pi / 6, 0, -np.pi / 6]  # 30°, 0°, -30°

# Create figure with two rows
fig, axes = plt.subplots(2, len(angles), figsize=(15, 8))

for col, angle in enumerate(angles):
    wave_data = WaveData(
        azimuth=angle, elevation=angle
    )  # Same angle for both directions

    # Plot elevation plane (top row)
    im_elev = axes[0, col].imshow(
        combined_apodization(
            sender=transmit_element,
            point_position=xyz_scan.get_points(),
            receiver=transmit_element,
            wave_data=wave_data,
        )
        .reshape(100, 100, 100)[49, :, :]
        .T,
        extent=[y_points[0], y_points[-1], z_points[-1], z_points[0]],
        aspect="auto",
        vmin=0,
        vmax=1,
    )
    axes[0, col].set_xlabel("y [m]")
    if col == 0:
        axes[0, col].set_ylabel("z [m]")
    axes[0, col].set_title(f"Angle: {angle*180/np.pi:.1f}° (y-z plane)")
    plt.colorbar(im_elev, ax=axes[0, col])

    # Plot azimuth plane (bottom row)
    im_azim = axes[1, col].imshow(
        combined_apodization(
            sender=transmit_element,
            point_position=xyz_scan.get_points(),
            receiver=transmit_element,
            wave_data=wave_data,
        )
        .reshape(100, 100, 100)[:, 49, :]
        .T,
        extent=[x_points[0], x_points[-1], z_points[-1], z_points[0]],
        aspect="auto",
        vmin=0,
        vmax=1,
    )
    axes[1, col].set_xlabel("x [m]")
    if col == 0:
        axes[1, col].set_ylabel("z [m]")
    axes[1, col].set_title(f"Angle: {angle*180/np.pi:.1f}° (x-z plane)")
    plt.colorbar(im_azim, ax=axes[1, col])

plt.suptitle("Combined Apodization: Elevation (top) vs Azimuth (bottom)")
plt.tight_layout()
plt.savefig("combined_apodization_2d.png")

# Calculate 3D volume for one specific angle (e.g., 30 degrees)
angle = np.pi / 6  # 30 degrees
wave_data = WaveData(azimuth=angle, elevation=angle)

# Calculate apodization values for full 3D volume
apod_3d = combined_apodization(
    sender=transmit_element,
    point_position=xyz_scan.get_points(),
    receiver=transmit_element,
    wave_data=wave_data,
).reshape(100, 100, 100)

# Create and save NIfTI file
# Note: Converting to RAS+ coordinate system for standard neuroimaging tools
apod_3d = np.transpose(apod_3d, (1, 0, 2))  # Swap x and y for RAS+ orientation
affine = np.diag(
    [
        (x_points[1] - x_points[0]) * 1000,  # Convert to mm
        (y_points[1] - y_points[0]) * 1000,
        (z_points[1] - z_points[0]) * 1000,
        1,
    ]
)
nifti_img = nib.Nifti1Image(apod_3d, affine)
nib.save(nifti_img, f"combined_apodization_3d_{angle*180/np.pi:.0f}deg.nii.gz")

plt.show()
```

which outputs the following apodization mask (note that the depth-visualization is flipped so that 0 is at the bottom):

image

**Comment by magnusdk** [comment]
Hi, thanks for picking this up :) I have a few suggestions:
1. I think we could probably have `PlaneWaveTransmitApodization` work in both azimuth and elevation at once by default instead of needing to combine two apodizations.
2. I think we should move away from using `vbeam.util.geometry.v2` until/unless it is properly finished. Having to call `ensure_2d_point` adds noise to the code. I take responsibility for this 😅

By the way, we are working towards a version 2 of vbeam which will simplify a lot of stuff. For example, we are adding a Probe object which means that we won't have to pass `array_bounds` into `PlaneWaveTransmitApodization` anymore. It will also use arrays with named dimensions (like xarray) instead of having a separate Spec object. This means that we can remove the ForAll(...) and Apply(..., Axis(...)) API as well, which is very verbose and complicated. I figured I'd give a heads up about this :) Work is being done on the `wip_spekk2.0.0` branch. I'd gladly help you migrate to the new version once it is out.

**Comment by magnusdk** [comment]
Here is a suggestion for an implementation that works in azimuth and elevation simultaneously:

```python
from typing import Optional, Tuple, Union

import jax
import jax.numpy as np
import matplotlib.pyplot as plt

from vbeam.apodization import Hanning, Rectangular, Window
from vbeam.core import Apodization, ElementGeometry, WaveData
from vbeam.fastmath.traceable import traceable_dataclass
from vbeam.util.geometry.v2 import distance


@traceable_dataclass(("array_bounds", "window"))
class PlaneWaveTransmitApodization(Apodization):
    # NOTE: This can be removed when we have the Probe object in vbeam 2.0
    array_bounds: Union[
        Tuple[np.ndarray, np.ndarray],  # 1D array
        Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray],  # 2D array
    ]
    window: Optional[Window] = None

    def __call__(
        self,
        sender: ElementGeometry,
        point_position: np.ndarray,
        receiver: ElementGeometry,
        wave_data: WaveData,
    ) -> float:
        # Normalized vector pointing in the direction of the transmitted wave
        direction = np.array(
            [
                np.sin(wave_data.azimuth) * np.cos(wave_data.elevation),
                np.sin(wave_data.elevation),
                np.cos(wave_data.azimuth) * np.cos(wave_data.elevation),
            ]
        )

        # Project the point onto the xy-plane with origin at the beam at depth z
        point_position = point_position - sender.position
        z = point_position[2]
        x = point_position[0] - direction[0] * z / direction[2]
        y = point_position[1] - direction[1] * z / direction[2]

        # Apply the window over the projected aperture and evaluate point
        window = self.window if self.window is not None else Rectangular()
        array_width = distance(self.array_bounds[1] - self.array_bounds[0])
        weight = window(np.abs(x / array_width))
        if len(self.array_bounds) == 4:  # If we use a 2D probe, also apply in elevation
            array_height = distance(self.array_bounds[3] - self.array_bounds[2])
            weight *= window(np.abs(y / array_height))
        return weight

# Example usage + plotting
point_position = np.stack(
    np.meshgrid(
        np.linspace(-0.1, 0.1, 100),
        np.linspace(-0.1, 0.1, 100),
        np.linspace(0, 0.2, 100),
        indexing="ij",
    ),
    axis=-1,
)
apod = PlaneWaveTransmitApodization(
    (
        np.array([-0.025, 0, 0]),
        np.array([0.025, 0, 0]),
        np.array([0, -0.01, 0]),
        np.array([0, 0.01, 0]),
    ),
    Hanning(),
)
# We also don't need to do this in vbeam 2.0 :) Named arrays makes broadcasting simple
apod = jax.vmap(apod, [None, 0, None, None])
apod = jax.vmap(apod, [None, 0, None, None])
apod = jax.vmap(apod, [None, 0, None, None])
result = apod(
    ElementGeometry(np.array([0, 0, 0])),
    point_position,
    None,
    WaveData(azimuth=0, elevation=0.1),
)
fig, ax = plt.subplots(ncols=2, figsize=(8, 4))
ax[0].imshow(result[:, 50, :].T, extent=(-0.1, 0.1, 0.2, 0.0))
ax[0].set_title("xz-plane")
ax[0].set_xlabel("x")
ax[0].set_ylabel("z")
ax[1].imshow(result[50, :, :].T, extent=(-0.1, 0.1, 0.2, 0.0))
ax[1].set_title("yz-plane")
ax[1].set_xlabel("y")
ax[1].set_ylabel("z")
fig.tight_layout()
```

**Comment by charlesbmi** [comment]
Thanks! Great news on `v2` ; happy to try a beta version if you want as well.

And thanks for implementing this 3D version in a nice way; I'll integrate your code into the PR.

**Comment by charlesbmi** [comment]
Updated this with your code suggestion @magnusdk and a couple input-validation checks. Feel free to add any changes or let me know if anything should be changed, thanks!

**Comment by magnusdk** [comment]
Great job! Thanks a lot :)

**Comment by charlesbmi** [comment]
Thanks @magnusdk ! could you also help run a release to `PyPi` ? (I bumped the `setup.py` version in this PR)

**Comment by magnusdk** [comment]
Yup, sorry, should be out now!

**Comment by charlesbmi** [comment]
Thanks!
